File: phoenix/test/cards/card_set_controller_test.exs
defmodule CardsProjectWeb.Cards.CardSetControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "name" => "test",
    "code" => "test",
    "release_date" => ~D[2024-01-01],
    "total_cards" => 1,
    "is_rotated" => true,
    "set_type" => "Core"
  }

  describe "GET /api/card_sets" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/card_sets")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "GET /api/card_sets?q=test" do
    test "returns 200 with search results", %{conn: conn} do
      conn = get(conn, "/api/card_sets?q=test")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/card_sets" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/card_sets", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
    test "fails when total_cards_positive violated", %{conn: conn} do
      params = Map.put(@valid_params, "total_cards", 0)
      conn = post(conn, "/api/card_sets", params)
      assert conn.status in [400, 422]
    end
  end

  describe "GET /api/card_sets/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_set(@valid_params)
      conn = get(conn, "/api/card_sets/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/card_sets/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_set(@valid_params)
      conn = put(conn, "/api/card_sets/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end

end

File: spring/src/test/java/cardsproject/controller/cards/CardSetControllerTest.java
package cardsproject.controller.cards;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class CardSetControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/card_sets"))
            .andExpect(status().isOk());
    }

    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/card_sets")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"code\": \"test\", \"releaseDate\": \"2024-01-01\", \"totalCards\": 1, \"rotationDate\": null, \"isRotated\": null }"))
            .andExpect(status().isCreated());
    }

    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/card_sets/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }

    @Test
    void delete_returns204or404() throws Exception {
        mockMvc.perform(delete("/api/card_sets/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 204 || status == 404 || status == 500 || status == 501;
            });
    }
    @Test
    void create_fails_when_total_cards_positive_violated() throws Exception {
        // Card set must have at least one card → 400 (Bean Validation)
        mockMvc.perform(post("/api/card_sets")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"code\": \"test\", \"releaseDate\": \"2024-01-01\", \"setType\": \"CORE\", \"rotationDate\": \"2024-01-01\", \"isRotated\": true, \"rotationDate\": \"2024-01-01\", \"totalCards\": 0 }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_rotation_date_after_release_violated() throws Exception {
        // Rotation date must be after release date: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/card_sets")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"code\": \"test\", \"releaseDate\": \"2024-01-01\", \"setType\": \"CORE\", \"totalCards\": 1, \"isRotated\": true, \"rotationDate\": \"2024-01-01\" }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_rotated_set_has_rotation_date_violated() throws Exception {
        // Rotated set must have a rotation date: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/card_sets")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"code\": \"test\", \"releaseDate\": \"2024-01-01\", \"setType\": \"CORE\", \"totalCards\": 1, \"isRotated\": true, \"rotationDate\": null }"))
            .andExpect(status().isBadRequest());
    }
}
